### PR TITLE
Deserialization constructors

### LINKED ISSUE #20887: Breaking change: non-public parameterless ctors will not be used for deserialization with JsonSerializer
### Version introduced

.NET 5.0 preview 1

### Old behavior

- On .NET Core 3.0+, we allow internal/private ctors of POCOs and call them while deserializing.
- On netstandard2.0/2.1, we don’t and throw a MissingMethodException: No parameterless constructor defined for this object. (from Activator.CreateInstance).

### New behavior

Non-public constructors are ignored by the serializer by default. This includes parameterless constructors. The serializer will use one of the following constructors for deserialization.

- Public constructor annotated with `JsonConstructorAttribute`.
- Public parameterless constructor.
- Public parameterized constructor (if it is the only public constructor present)

If none of the above are present, a `NotSupportedException` will be thrown if deserialization of the type is attempted.

### Reason for change

- Enforce consistent behavior between all TFMs that System.Text.Json builds for (.NET Core 3.0+, .netstandard 2.0+)
- JsonSerializer shouldn't be calling non-public surface area of types by default (whether that's ctor, properties, or fields).

### Recommended action

- If the user controls the own type and it is feasible, make the parameterless constructor public
- Implement a `JsonConverter<T>` for the type and control the deserialization behavior

### Category

- [ ] ASP.NET Core
- [ ] C#
- [ ] Code analysis
- [x] Core .NET libraries
- [ ] Cryptography
- [ ] Data
- [ ] Debugger
- [ ] Deployment
- [ ] Globalization
- [ ] Interop
- [ ] JIT
- [ ] LINQ
- [ ] Managed Extensibility Framework (MEF)
- [ ] MSBuild
- [ ] Networking
- [ ] Printing
- [ ] Security
- [x] Serialization
- [ ] Visual Basic
- [ ] Windows Forms
- [ ] Windows Presentation Foundation (WPF)
- [ ] XML, XSLT

### Affected APIs

[`JsonSerializer.Deserialize`](https://docs.microsoft.com/dotnet/api/system.text.json.jsonserializer.deserialize?view=netcore-3.1)
[`JsonSerializer.DeserializeAsync`](https://docs.microsoft.com/dotnet/api/system.text.json.jsonserializer.deserializeasync?view=netcore-3.1)

---
#### Issue metadata

* Issue type: breaking-change

### PR DESCRIPTION
**Pull request by gewarren**
Fixes #20887.

[Preview link](https://review.docs.microsoft.com/en-us/dotnet/core/compatibility/serialization?branch=pr-en-us-21025#non-public-parameterless-constructors-not-used-for-deserialization).

### CODE REVIEW

File: includes/core-changes/serialization/5.0/non-public-parameterless-constructors-not-used-for-deserialization.md [hunk 1]
```diff
@@ -0,0 +1,47 @@
...
+
+If none of these constructors are available, a <xref:System.NotSupportedException> is thrown if you attempt to deserialize the type.
+
+#### Version introduced
+
+5.0
+
+#### Reason for change
+
+- To enforce consistent behavior between all target framework monikers (TFMs) that <xref:System.Text.Json?displayProperty=fullName> builds for (.NET Core 3.0 and later version, and .NET Standard 2.0 and 2.1)
```
Comment (tdykstra) [inline includes/core-changes/serialization/5.0/non-public-parameterless-constructors-not-used-for-deserialization.md:23]:
```suggestion
- To enforce consistent behavior between all target framework monikers (TFMs) that <xref:System.Text.Json?displayProperty=fullName> builds for (.NET Core 3.0 and later versions, and .NET Standard 2.0 and 2.1)
```

**Review by tdykstra — APPROVED** [review]
LGTM, one suggestion.

File: includes/core-changes/serialization/5.0/non-public-parameterless-constructors-not-used-for-deserialization.md [hunk 2]
```diff
@@ -0,0 +1,47 @@
+### Non-public, parameterless constructors not used for deserialization
+
+For consistency across all supported target framework monikers (TFMs), non-public, parameterless constructors are no longer used for deserialization with <xref:System.Text.Json.JsonSerializer>, by default.
+
+#### Change description
+
+On .NET Core 3.0 and 3.1, internal and private constructors can be used for deserialization. On .NET Standard 2.0 and 2.1, internal and private constructors are not allowed, and a <xref:System.MissingMethodException> is thrown if no public, parameterless constructor is defined.
```
Comment (ahsonkhan) [inline includes/core-changes/serialization/5.0/non-public-parameterless-constructors-not-used-for-deserialization.md:7]: Netstandard isn't really a TFM, so it would be better to rephrase this to say something like: the standalone NuGet packages version `4.6.0-4.7.2` that target netstandard 2.0 (and higher) behaved one way (inconsistently) ... and .NET 5.0/NuGet package 5.0.0 (and higher) behave consistently.

https://www.nuget.org/packages/System.Text.Json/

- .NET Core 3.0 release has STJ built-in and we also shipped a 4.6.0 standalone STJ NuGet package that supports ns2.0 to support older platform.
- .NET Core 3.1 release along with 4.7.0 STJ NuGet package (which continues to support ns2.0).
- .NET Core 5.0 release along with 5.0.0 STJ NuGet package (which continues to support ns2.0).

File: includes/core-changes/serialization/5.0/non-public-parameterless-constructors-not-used-for-deserialization.md [hunk 3]
```diff
@@ -0,0 +1,47 @@
...
+
+If none of these constructors are available, a <xref:System.NotSupportedException> is thrown if you attempt to deserialize the type.
+
+#### Version introduced
+
+5.0
+
+#### Reason for change
+
+- To enforce consistent behavior between all target framework monikers (TFMs) that <xref:System.Text.Json?displayProperty=fullName> builds for (.NET Core 3.0 and later versions, and .NET Standard 2.0 and 2.1)
```
Comment (ahsonkhan) [inline includes/core-changes/serialization/5.0/non-public-parameterless-constructors-not-used-for-deserialization.md:23]: Stating ns2.1 isn't necessary. NS is concentric circle, anything that supports n2.0 automatically supports ns2.1 and any future versions, if any.

File: includes/core-changes/serialization/5.0/non-public-parameterless-constructors-not-used-for-deserialization.md [hunk 4]
```diff
@@ -0,0 +1,47 @@
...
+- If you own the type and it's feasible, make the parameterless constructor public.
+- Otherwise, implement a `JsonConverter<T>` for the type and control the deserialization behavior.
+
+#### Category
+
+Serialization
+
+#### Affected APIs
+
+- <xref:System.Text.Json.JsonSerializer.Deserialize%2A?displayProperty=fullName>
```
Comment (ahsonkhan) [inline includes/core-changes/serialization/5.0/non-public-parameterless-constructors-not-used-for-deserialization.md:37]: Should we list all overloads of these APIs?

File: includes/core-changes/serialization/5.0/non-public-parameterless-constructors-not-used-for-deserialization.md [hunk 5]
```diff
@@ -0,0 +1,47 @@
...
+- <xref:System.Text.Json.JsonSerializer.DeserializeAsync%2A?displayProperty=fullName>
+
+<!--
+
+#### Affected APIs
+
+- `Overload:System.Text.Json.JsonSerializer.Deserialize`
+- `Overload:System.Text.Json.JsonSerializer.DeserializeAsync`
+
+-->
```
Comment (ahsonkhan) [inline includes/core-changes/serialization/5.0/non-public-parameterless-constructors-not-used-for-deserialization.md:47]: Why do we have this commented out section?

Comment (gewarren) [reply includes/core-changes/serialization/5.0/non-public-parameterless-constructors-not-used-for-deserialization.md:47]: It's used by an analyzer.

File: includes/core-changes/serialization/5.0/non-public-parameterless-constructors-not-used-for-deserialization.md (hunk 4 above)
Comment (gewarren) [reply includes/core-changes/serialization/5.0/non-public-parameterless-constructors-not-used-for-deserialization.md:37]: If it applies to all overloads, we just specify it this way.

File: includes/core-changes/serialization/5.0/non-public-parameterless-constructors-not-used-for-deserialization.md (hunk 2 above)
Comment (ahsonkhan) [reply includes/core-changes/serialization/5.0/non-public-parameterless-constructors-not-used-for-deserialization.md:7]: cc @layomia please help @gewarren update the wording, if you agree :)
